### ISSUE TITLE
Do not display fingerprint unlock option right after startup

### ISSUE THREAD

**Issue by MPeti1** [beemdevelopment/Aegis#550]
Currently (1.2.1) if fingerprint unlock is set up, when you open the app it'll instantly give it away that the database can be unlocked with biometrics.
This can be problematic, because if you are forced you can't say that you don't know the password. Also there are countries where it's not allowed for police officers to force you to input your password, but they are not denied from holding your hand to the fingerprint reader ([source](https://www.wired.com/story/police-unlock-iphone-face-id-legal-rights/))

I would like to request to not automatically open the biometrics popup if biometrics unlock is configured.
Also, if you currently only show the "Or tap here to use biometrics" text if biometrics unlock is configured, I would like to request to make it unconditional, so it's always shown, with the difference that the popup would never accept the fingerprint.

While this does not solve the actual problem, that with forcing you the app could be unlocked, this could help you persuade the other party that biometrics unlock is not set up (e.g. by setting it up for an other finger on your hand, or for a finger on your toe, which nobody would think to try).

**Comment by alexbakker** [comment]
I think you're trying to solve this issue at the wrong level. Do you know about Android's lockdown mode? If so, does that not suffice? I'd argue that's your only chance of convincing the authorities that you don't have biometrics configured, by hiding that fact at the lockscreen level, rather than at the application level.

If I understood you correctly, there also seems to be some misunderstanding about how biometric authentication works on Android. Apps can only prompt for biometrics if you've actually set it up in Android settings. Once shown, apps have no control over that prompt, as the system handles the authentication transparently. There is also no way for apps to know which finger you used to dismiss the biometrics prompt. Any of the enrolled fingerprints (or faces) can be used to dismiss it.

Obviously, if the authorities manage to unlock your phone, they would be able to check the settings in Android and see that you have biometrics configured. The jig is up if Aegis then also prompts you for biometrics.

**Comment by MPeti1** [comment]
> Do you know about Android's lockdown mode? If so, does that not suffice?

I didn't know about it, but my phone (Android 6) doesn't support it.

> Once shown, apps have no control over that prompt, as the system handles the authentication transparently.

My point would be to not immediately show the prompt when the app is opened, but only after clicking "Or tap here to use biometrics"

> Apps can only prompt for biometrics if you've actually set it up in Android settings.

I understand that. It's because it's based on what fingers/faces/irises I've set up in the system, and because it's global across apps, and because this way apps don't have access to this sensitive information

> Once shown, apps have no control over that prompt, as the system handles the authentication transparently.

I understand that

> There is also no way for apps to know which finger you used to dismiss the biometrics prompt.

I understand that too, and I didn't want to imply that it would be the task of your app.
What I wanted to mean here is that the user could just set up a finger that's not obvious to try (maybe with some rotation so it's easy to say that none of the fingers are configured) and in the end the popup in front of the app would just say "not recognized" at all times.

> Any of the enrolled fingerprints (or faces) can be used to dismiss it.

So apps can't request authentication to be done with 2 different types of biometrics methods? I didn't know that, but good to know.

> Obviously, if the authorities manage to unlock your phone, they would be able to check the settings in Android and see that you have biometrics configured.

But does that imply that Aegis would request biometrics based authentication? On my phone there is no such menu that would show which applications use this feature

**Comment by alexbakker** [comment]
Overall, I think the changes proposed here would add inconvenience and confusion for regular users that are not worth the questionable security benefits claimed here. Especially since more recent versions of Android have a feature that solves your concerns.

> I didn't know about it, but my phone (Android 6) doesn't support it.

Regardless of whether your old device supports it or not, would you agree that Android's lockdown mode would suffice? To be frank, I would be more concerned about running such an old Android version, than the authorities compelling me to use biometrics to unlock my phone.

> So apps can't request authentication to be done with 2 different types of biometrics methods?

I'm not sure what you mean by that. But let's say you set up biometric authentication in Aegis with finger 1. Any other fingers registered on the device will also be able to unlock Aegis.

> But does that imply that Aegis would request biometrics based authentication? On my phone there is no such menu that would show which applications use this feature

No, that was just a general statement.

**Comment by MPeti1** [comment]
> Overall, I think the changes proposed here would add inconvenience and confusion for regular users that are not worth the questionable security benefits claimed here. Especially since more recent versions of Android have a feature that solves your concerns.

I don't think it would add inconvenience or confusion. The button is already there to unlock the app with biometrics authentication (it's directly below the password field), and also it's almost pointless to have that button as of now, because it has no uses.

> Regardless of whether your old device supports it or not, would you agree that Android's lockdown mode would suffice?

It seems to me that in case lockdown mode is turned on, it suffices, though it's easy to forget that you need to turn it on again after unlocking your phone. I think it would be much better if the auto-disabling of it could be turned off, but we can't really change that

> To be frank, I would be more concerned about running such an old Android version, than the authorities compelling me to use biometrics to unlock my phone.

I'm aware of that, and I'm not asking for this change because I think that it might help specifically me.

> I'm not sure what you mean by that. But let's say you set up biometric authentication in Aegis with finger 1. Any other fingers registered on the device will also be able to unlock Aegis.

There are phones that can do biometric authentication through both fingerprint reading, iris scanning, and face recognition, because they have the required hardware for it. They usually allow the user to make unlocking the lock screen require at least 2 of these: fingerprint, iris, face, input (pin/pattern/password)

**Comment by alexbakker** [comment]
> I don't think it would add inconvenience or confusion. The button is already there to unlock the app with biometrics authentication (it's directly below the password field), and also it's almost pointless to have that button as of now, because it has no uses.

We actually have some evidence that this would be an inconvenience to users. Users who use biometrics to unlock Aegis get a password reminder every now and then. When the password reminder is shown, the keyboard pops up instead of the biometrics prompt. We've received reports from users who complained that they had to manually press the biometrics button every time to unlock the app. With this proposal also requiring to show the biometrics button without even having configured it in Aegis will only add to the confusion, I'm afraid. Of course, if there were convincing security benefits here, we could decide that some inconvenience is acceptable.

The button is actually not pointless. It's there so that users who accidentally dismiss the biometrics prompt can get it back. It's also there so that users who realize they have forgotten their password after the password reminder is shown can still unlock the app.

> I'm aware of that, and I'm not asking for this change because I think that it might help specifically me.

The point I was trying to make is that it would make more sense for users who are as security conscious as you to prioritize getting a device with a newer version of Android, in my opinion. As a nice bonus, the lockdown mode will also become available then.

> There are phones that can do biometric authentication through both fingerprint reading, iris scanning, and face recognition, because they have the required hardware for it. They usually allow the user to make unlocking the lock screen require at least 2 of these: fingerprint, iris, face, input (pin/pattern/password)

Out of curiosity, which devices have an option to require multiple factors to unlock? Apps cannot request a specific form of biometric authentication, just biometric authentication in general.

**Comment by MPeti1** [comment]
> We've received reports from users who complained that they had to manually press the biometrics button every time to unlock the app.

Oh. I see. So this wasn't the original idea, but people wanted it.
Please don't take the following as an insult or something, because it really isn't against you, but now I see why Android nowadays aims to be more and more simple, even with the cost of having less useful features. If simple people start using a software, they will actually have requests to oversimplify things for them to be easier, and this is what happened to you too, you had to adapt to the users who are not willing to press a single button to avoid pressing multiple buttons.

I always wondered why aren't they making settings for these, instead of just taking out the features?
I mean they could hide it as the developer settings, but that might be too much.
They could make a similar oversimplified UI as the "[lite mode](https://youtu.be/tCyrpDHelqU)" on some phones, just with a little more features that the average people use. This would be default on, and it could be turned off in the settings.
Or they could just place settings in the Settings app, to turn specific features on or off, as mainstream people never look there anyways, so it won't be confusing for them.
Of course I haven't found the occasion to ask an Android OS developer, so I couldn't ask these questions, but I think you might have an answer for the 3rd solution: why making settings to toggle features/behavior is more rare, then just taking out features? for example in your app, would it have a disadvantage for making this a toggleable setting?

> The point I was trying to make is that it would make more sense for users who are as security conscious as you to prioritize getting a device with a newer version of Android, in my opinion.

You're totally right, and it's being worked on to have a more up-to-date phone. I'm researching which one would be the best for me, but still affordable (I'm a student), and I try to find and avoid traps that I would otherwise only find after I've already spent my money.

> Out of curiosity, which devices have an option to require multiple factors to unlock?

Honestly I don't know examples right now, but I'm sure that I've both read about such devices, and I think a friend of me had this sen up on their phone. However I did a quick search, and now I really can't find anything that would allow requiring both biometric and input for unlock. I thought at least the lock screen can be protected this way

**Comment by alexbakker** [comment]
> Please don't take the following as an insult or something

Not at all!

> why making settings to toggle features/behavior is more rare, then just taking out features? for example in your app, would it have a disadvantage for making this a toggleable setting?

It really just comes down to app developers having to make choices. We can't implement every possible feature, as each one we add increases the maintenance burden. Instead, we try to set sane defaults and offer customizability only where necessary. We accept proposals for features that make sense to us, are useful for more than just a handful of users, are in-scope and maintainable.

In some area's, like security, we're extra strict. As you're probably noticed, there's not a lot to customize in the security settings of your vault. You either choose no security or great security. There's nothing in between, and there is no cryptographic agility. Of course, we don't have to be as strict for UX features, but the criteria above still apply.
